src/gerbil/runtime/util.ss

```diff
@@ -518,29 +518,37 @@ namespace: #f
 
 (def (display-as-string x port)
   (cond
-   ((or (string? x) (symbol? x) (keyword? x) (number? x) (char? x)) (display x port))
-   ((pair? x) (display-as-string (car x) port) (display-as-string (cdr x) port))
-   ((vector? x) (display-as-string (vector->list x) port))
-   ((or (null? x) (void? x) (eof-object? x) (boolean? x)) (void))
+   ((or (string? x) (symbol? x) (keyword? x) (number? x) (char? x))
+    (display x port))
+   ((pair? x)
+    (display-as-string (car x) port)
+    (display-as-string (cdr x) port))
+   ((vector? x)
+    (vector-for-each (cut display-as-string <> port) x))
+   ((or (null? x) (void? x) (eof-object? x) (boolean? x))
+    (void))
    (else (error "cannot convert to symbol" x))))
 
-(def as-string
-  (case-lambda
-    ((x) (cond ((string? x) x)
-               ((symbol? x) (symbol->string x))
-               ((keyword? x) (keyword->string x))
-               (else (call-with-output-string [] (lambda (p) (display-as-string x p))))))
-    (r (as-string r))))
-
-(def make-symbol
-  (case-lambda
-    ((x) (if (interned-symbol? x) x (string->symbol (as-string x))))
-    (r (make-symbol r))))
-
-(def make-keyword
-  (case-lambda
-    ((x) (if (interned-keyword? x) x (string->keyword (as-string x))))
-    (r (make-keyword r))))
+(def* as-string
+  ((x)
+   (cond
+    ((string? x) x)
+    ((symbol? x)
+     (symbol->string x))
+    ((keyword? x)
+     (keyword->string x))
+    (else
+     (call-with-output-string [] (cut display-as-string x <>)))))
```

Comment:
Will gambit be able to de-duplicate? 

Should we use this, and so on afterwards?
```
(def as-string
  (let (use-display
         (lambda (x) (call-with-output-string [] (cut display-as-string x <>))))
    (case-lambda
      ((x)
        (cond
          ((string? x) x)
          ((symbol? x)
           (symbol->string x))
          ((keyword? x)
           (keyword->string x))
          (else
           (use-display x))))
      (args
       (use-display args)))))
```

Reply:
it will inline in that case resulting in the same code.

Reply:
also you should almost never let over case/opt/kw lamda, the compiler wont be ablr to direct dispatch.